### PR TITLE
Allow the OverlapDetector to be generic over input feature types

### PR DESCRIPTION
**Pull request by clintval**
I'm interested in seeing what others think given that the OverlapDetector isn't an ergonomic utility to use, especially for custom interval-types which I may often choose to implement myself, or have received from a third-party package/API.

## Goals

1. The ergonomics of the OverlapDetector could be improved to store and query **_generic_** intervals.
2. This PR should introduce the fewest backwards incompatibilities

## Example with Existing API

The following demonstrates using a custom interval type with the OverlapDetector:

```python
@dataclass(eq=True, frozen=True)
class CustomPrimer:
    """A custom primer feature in the shape of a BED3+1 record."""

    refname: str
    start: int
    end: int
    primary_key: str

panel: list[CustomPrimer] = [
    CustomPrimer("chr1", 1, 4, "primer1"),
    CustomPrimer("chr1", 5, 9, "primer2"),
]

detector = OverlapDetector()
for primer in panel:
    detector.add(Interval(primer.chrom, primer.start, primer.end, name=primer.primary_key))

primer_lookup: dict[str, CustomPrimer] = {primer.primary_key: primer for primer in panel}
overlapping_intervals: list[Interval] = detector.get_overlaps(Interval("chr1", 2, 3))
overlapping_primers: list[CustomPrimer] = [primer_lookup[interval.name] for interval in overlapping_intervals]
```

What makes this challenging to use is:

1. The `OverlapDetector` only supports storing [`Interval`](https://github.com/fulcrumgenomics/pybedlite/blob/3a4b23a7fd8369edb810e90f4c19a5b763a0f561/pybedlite/overlap_detector.py#L129) objects
2. The `OverlapDetector` can only be queried with [`Interval`](https://github.com/fulcrumgenomics/pybedlite/blob/3a4b23a7fd8369edb810e90f4c19a5b763a0f561/pybedlite/overlap_detector.py#L129) objects
4. This library's primary data structure [`BedRecord`](https://github.com/fulcrumgenomics/pybedlite/blob/3a4b23a7fd8369edb810e90f4c19a5b763a0f561/pybedlite/bed_record.py#L47) doesn't work with `OverlapDetector`
    - Often requires the user to call a [`to_interval()`](https://github.com/fulcrumgenomics/pybedlite/blob/3a4b23a7fd8369edb810e90f4c19a5b763a0f561/pybedlite/bed_record.py#L206-L228) converter
    - Often requires the user to call a [`from_bedrecord()`](https://github.com/fulcrumgenomics/pybedlite/blob/3a4b23a7fd8369edb810e90f4c19a5b763a0f561/pybedlite/overlap_detector.py#L105-L126) converter
5. In order to convert back to your data of interest, you must knowingly store a primary key in the `name` field of `Interval` and perform a dictionary lookup (as shown above). If you don't have a primary key, you will need to make one.
6. Any intervals must be added separately from class instantiation, often adding 2 LOC.

## Example with this PR

```python
@dataclass(eq=True, frozen=True)
class CustomPrimer:
    """A custom primer feature in the shape of a BED3+1 record."""

    refname: str
    start: int
    end: int
    primary_key: str

panel: list[CustomPrimer] = [
    CustomPrimer("chr1", 1, 4, "primer1"),
    CustomPrimer("chr1", 5, 9, "primer2"),
]

detector: OverlapDetector[CustomPrimer] = OverlapDetector(panel)
overlapping_primers: list[CustomPrimer] = detector.get_overlaps(Interval("chr1", 2, 3))
```

*NB: although we are using an `Interval` to query the OverlapDetector, the type system is aware that we built the OverlapDetector with `CustomPrimer` types and all overlap functions will always return `CustomPrimer` objects. A user could alternatively query the overlap detector with any other interval-like feature but still receive `CustomerPrimer` objects which overlap.*

*NB: the BED spec notation BED3+1 is described [here](https://github.com/samtools/hts-specs/blob/master/BEDv1.pdf)*

### CODE REVIEW

File: pybedlite/overlap_detector.py [hunk 1]
```diff
@@ -56,6 +61,52 @@
...
+    def contig(self) -> str:
+        """A reference sequence name."""
+
+
+class _GenomicSpanWithRefName(_Span, Hashable, Protocol):
+    """A genomic feature where reference sequence is accessed with `refname`."""
+
+    @property
+    def refname(self) -> str:
+        """A reference sequence name."""
```
Comment (clintval) [inline pybedlite/overlap_detector.py:97]: These are private [structural types](https://docs.python.org/3/library/typing.html#typing.Protocol). I intentionally did not make them public or [runtime checkable](https://docs.python.org/3/library/typing.html#typing.runtime_checkable).

File: pybedlite/overlap_detector.py [hunk 2]
```diff
@@ -126,36 +177,94 @@ def from_bedrecord(cls: Type["Interval"], record: BedRecord) -> "Interval":
...
-class OverlapDetector(Iterable[Interval]):
+_GenericGenomicSpan = TypeVar(
+    "_GenericGenomicSpan",
+    bound=Union[_GenomicSpanWithChrom, _GenomicSpanWithContig, _GenomicSpanWithRefName],
+)
+"""
+A generic genomic feature where the reference sequence name is accessed with any of the 3 most
+common property names ("chrom", "contig", "refname"). This type variable is used for describing the
+generic type contained within the :class:`~pybedlite.overlap_detector.OverlapDetector`.
+"""
```
Comment (clintval) [inline pybedlite/overlap_detector.py:188]: This redefinition of `GenomicSpan` is necessary to decouple the generic type parameter of the overlap detector with the types it will receive in any of it's overlap methods.

File: pybedlite/overlap_detector.py [hunk 3]
```diff
@@ -126,36 +177,94 @@ def from_bedrecord(cls: Type["Interval"], record: BedRecord) -> "Interval":
...
 
-    def __init__(self) -> None:
+    def __init__(self, intervals: Optional[Iterable[_GenericGenomicSpan]] = None) -> None:
         # A mapping from the contig/chromosome name to the associated interval tree
         self._refname_to_tree: Dict[str, cr.cgranges] = {}  # type: ignore
         self._refname_to_indexed: Dict[str, bool] = {}
-        self._refname_to_intervals: Dict[str, List[Interval]] = {}
+        self._refname_to_intervals: Dict[str, List[_GenericGenomicSpan]] = {}
+        if intervals is not None:
+            self.add_all(intervals)
```
Comment (clintval) [inline pybedlite/overlap_detector.py:221]: Small feature addition that would be nice to have regardless of this PR.

File: pybedlite/overlap_detector.py [hunk 4]
```diff
@@ -126,36 +177,94 @@ def from_bedrecord(cls: Type["Interval"], record: BedRecord) -> "Interval":
...
+                hasattr(interval, "strand")
+                and isinstance(interval.strand, BedStrand)
+                and interval.strand is BedStrand.Negative
+            )
+            or (
+                hasattr(interval, "strand")
+                and isinstance(interval.strand, str)
+                and interval.strand == "-"
+            )
+        )
```
Comment (clintval) [inline pybedlite/overlap_detector.py:256]: Although these are pure functions and I could have implemented them in the package namespace, I made them class static methods because an advanced user (developer) may want to temporarily override them via inheritance.

File: pybedlite/overlap_detector.py [hunk 5]
```diff
@@ -126,36 +177,94 @@ def from_bedrecord(cls: Type["Interval"], record: BedRecord) -> "Interval":
...
+
+    def add(self, interval: _GenericGenomicSpan) -> None:
         """Adds an interval to this detector.
 
         Args:
             interval: the interval to add to this detector
         """
-        refname = interval.refname
+        if not isinstance(interval, Hashable):
+            raise ValueError(f"Interval feature is not hashable but should be: {interval}")
```
Comment (clintval) [inline pybedlite/overlap_detector.py:265]: Small feature addition that would be nice to have regardless of this PR.

This is necessary because in some class methods, `set` is called.

File: pybedlite/overlap_detector.py [hunk 6]
```diff
@@ -267,9 +384,9 @@ def from_bed(cls, path: Path) -> "OverlapDetector":
         Returns:
             An overlap detector for the regions in the BED file.
         """
-        detector = OverlapDetector()
+        detector: OverlapDetector[BedRecord] = OverlapDetector()
 
         for region in BedSource(path):
-            detector.add(Interval.from_bedrecord(region))
+            detector.add(region)
```
Comment (clintval) [inline pybedlite/overlap_detector.py:390]: No more converter! It is worth noting this is technically a backwards incompatible change since the returned OverlapDetector will no longer yield `Interval` objects, but now `BedRecord` objects.

File: pybedlite/tests/test_overlap_detector.py [hunk 7]
```diff
@@ -188,3 +193,61 @@ def test_construction_from_interval(bed_records: List[BedRecord]) -> None:
             assert new_record.strand is BedStrand.Positive
         else:
             assert new_record.strand is record.strand
+
+
+def test_arbitrary_interval_types() -> None:
```
Comment (clintval) [inline pybedlite/tests/test_overlap_detector.py:198]: I will need a few more tests if reviews are favorable.

But this 1 test shows a (_hopefully unrealistic_) complex example which flexes most functionality.

File: pybedlite/overlap_detector.py [hunk 8]
```diff
@@ -126,36 +177,94 @@ def from_bedrecord(cls: Type["Interval"], record: BedRecord) -> "Interval":
...
     The same interval may be added multiple times, but only a single instance will be returned
-    when querying for overlaps.  Intervals with the same coordinates but different names are
-    treated as different intervals.
+    when querying for overlaps.
 
     This detector is the most efficient when all intervals are added ahead of time.
     """
 
-    def __init__(self) -> None:
+    def __init__(self, intervals: Optional[Iterable[_GenericGenomicSpan]] = None) -> None:
```
Comment (nh13) [inline pybedlite/overlap_detector.py:215]: what about passing as `*args` instead?

**Review by nh13 — COMMENTED** [review]
I worry about _performance_ of this PR.  See #6 for an example of how we may want to test this.

File: pybedlite/overlap_detector.py (hunk 8 above)
Comment (clintval) [reply pybedlite/overlap_detector.py:215]: Could do that too! My gut feeling is users have a large collection of intervals to put into an overlap detector in the default case, so a list already exists when commonly building this class. So, instead of splatting the list into a varargs, one could use that list directly. I think it is less common that users will have just a few intervals not in any collection to be input into the OverlapDetector.

**Comment by clintval** [comment]
> I worry about _performance_ of this PR. See #6 for an example of how we may want to test this.

I'll find a way to test that! I don't think there will be any performance issues _**using**_ the OverlapDetector but it is possible building one will be a little slower since you need to check each object you add for a reference name property.

**Comment by nh13** [comment]
@clintval that's the case I'm worried about.  I wonder if we can avoid it somehow.

**Comment by tfenne** [comment]
@clintval I skimmed the PR briefly and wanted to suggest an alternative to see you have considered it and discarded it.  The OverlapDetector is named for the same construct from HTSJDK ... and that version does things slightly differently.  Beyond the fact that it can take any implementation of `Locatable` as the "Interval" type, it also decouples the intervals/spans from the returned element. 

At it's simplest you can think of it as:

```
overlap_detector.add(interval, primer)
primers = overlap_detector.get_overlaps(interval)
```

I.e. for every interval you associate the object you'd like returned when a query overlaps that interval.  It fixes some, but not all, of your issues I think - primarily that you don't need to keep an external mapping of interval back to your source type.

**Comment by clintval** [comment]
> @clintval I skimmed the PR briefly and wanted to suggest an alternative to see you have considered it and discarded it. The OverlapDetector is named for the same construct from HTSJDK ... and that version does things slightly differently. Beyond the fact that it can take any implementation of `Locatable` as the "Interval" type, it also decouples the intervals/spans from the returned element.
> 
> At it's simplest you can think of it as:
> 
> ```
> overlap_detector.add(interval, primer)
> primers = overlap_detector.get_overlaps(interval)
> ```
> 
> I.e. for every interval you associate the object you'd like returned when a query overlaps that interval. It fixes some, but not all, of your issues I think - primarily that you don't need to keep an external mapping of interval back to your source type.

@tfenne I thought about that too but didn't give it a lot of thought because for large interval collections it seems it could ~2x your memory footprint by having a sibling object for every object you actually want to query. Does that seem right?

**Comment by tfenne** [comment]
I think that's right @clintval - more memory usage for that solution vs. likely more CPU for having to indirect function calls.

**Comment by nh13** [comment]
@clintval another solution is to _sub-class_ `Interval` and add an attribute that you'd like to carry with that interval.  This would not require an "external map", as well as not bloat memory for folks that just want to store intervals.  I have successfully used this pattern elsewhere.

**Comment by clintval** [comment]
> @clintval another solution is to _sub-class_ `Interval` and add an attribute that you'd like to carry with that interval. This would not require an "external map", as well as not bloat memory for folks that just want to store intervals. I have successfully used this pattern elsewhere.

I like it! That is a good idea for how to use the existing classes and functions without the need for creating a dictionary. It would work well for custom genomic features. It would still require making an interval companion object for every genomic feature and wouldn't work for 3rd-party genomic features (where I cannot define the class hierarchy), but it's a little more elegant than what I've been doing.

File: pybedlite/overlap_detector.py [hunk 9]
```diff
@@ -126,36 +177,94 @@ def from_bedrecord(cls: Type["Interval"], record: BedRecord) -> "Interval":
...
-    def add(self, interval: Interval) -> None:
+    @staticmethod
+    def _reference_sequence_name(interval: GenomicSpan) -> str:
+        """Return the reference name of a given interval."""
+        if isinstance(interval, Interval) or hasattr(interval, "refname"):
+            return interval.refname
+        elif isinstance(interval, BedRecord) or hasattr(interval, "chrom"):
+            return interval.chrom
+        elif hasattr(interval, "contig"):
+            return interval.contig
```
Comment (nh13) [inline pybedlite/overlap_detector.py:235]: How about having the `_reference_sequence_name` method be an _instance_ method that must be implemented by all sub-classes with the above default implementation, and can be overriden by `Interval` and `BedRecord`:

```python
class Interval:
    ...
    def _reference_sequence_name((self) -> str:
        return self.refname

class BedRecord:
    ...
    def _reference_sequence_name((self) -> str:
        return self.chrom

**Comment by clintval** [comment]
Backwards incompatible changes summarized:

1. To appease _**MyPy only**_ users must now specify the inner type of the OverlapDetector, e.g.:

    ```python
    detector: OverlapDetector[BedRecord] = OverlapDetector()
    ````

2. The classmethod `OverlapDetector.from_bed()` now stores the `BedRecord`s themselves instead of Intervals:

    - Previous method and type signature: https://github.com/fulcrumgenomics/pybedlite/blob/592fb8e0753a3ef9da0bf8b97bf6cad904af2dea/pybedlite/overlap_detector.py#L314-L327
    - Current method and type signature: https://github.com/fulcrumgenomics/pybedlite/blob/01b09e2f74d6b7924bccb304428f48d5a9a39ddb/pybedlite/overlap_detector.py#L400-L413

**Comment by clintval** [comment]
There should be no performance regressions with this solution.

A necessary inconvenience was the addition of a duplicate getter and one helper on the `BedRecord` for these properties:

https://github.com/fulcrumgenomics/pybedlite/blob/01b09e2f74d6b7924bccb304428f48d5a9a39ddb/pybedlite/bed_record.py#L185-L193

Which are are derived from fields upon the class:

https://github.com/fulcrumgenomics/pybedlite/blob/01b09e2f74d6b7924bccb304428f48d5a9a39ddb/pybedlite/bed_record.py#L70-L75

File: .gitmodules [hunk 10]
```diff
@@ -1,3 +1,4 @@
 [submodule "cgranges"]
 	path = cgranges
 	url = https://github.com/lh3/cgranges
+	ignore = dirty
```
Comment (clintval) [inline .gitmodules:4]: Every time this submodule is compiled (`poetry install`) untracked files are added to the directory.

This tells Git to ignore those untracked compiled files.

File: pybedlite/overlap_detector.py [hunk 11]
```diff
@@ -178,55 +231,75 @@ def from_ucsc(
             ) from exception
 
 
-class OverlapDetector(Iterable[Interval]):
+GenericGenomicSpan = TypeVar("GenericGenomicSpan", bound=Union[GenomicSpan, StrandedGenomicSpan])
```
Comment (msto) [inline pybedlite/overlap_detector.py:234]:
```suggestion
S = TypeVar("S", bound=Union[GenomicSpan, StrandedGenomicSpan])
```

**issue** Could this be made a single uppercase letter?

I would ideally prefer if type variables/parameters could be named as one (or two, maximally) uppercase letters. I think this is more consistent with convention[^1][^2][^3], and improves the clarity of the code. It helps distinguish type parameters from other classes, and makes it more clear when a class is generic. 

[^1]: > By convention, type parameter names are single, uppercase letters. This stands in sharp contrast to the variable [naming](https://docs.oracle.com/javase%2Ftutorial%2F/java/nutsandbolts/variables.html#naming) conventions that you already know about, and with good reason: Without this convention, it would be difficult to tell the difference between a type variable and an ordinary class or interface name.
https://docs.oracle.com/javase%2Ftutorial%2F/java/generics/types.html

[^2]: > For simple type parameters, a single upper-case letter (from the English alphabet) should be used, starting with `A` (this is different than the Java convention of starting with `T`)
https://docs.scala-lang.org/style/naming-conventions.html#type-parameters-generics

[^3]: > To parameterize the types in a new single function, we need to name the type parameter, just as we do for the value parameters to a function. You can use any identifier as a type parameter name. But we’ll use T because, by convention, type parameter names in Rust are short, often just one letter, and Rust’s type-naming convention is UpperCamelCase. Short for type, T is the default choice of most Rust programmers.
https://doc.rust-lang.org/book/ch10-01-syntax.html

**Review by msto — CHANGES_REQUESTED** [review]
A few minor comments.

I think this represents a welcome improvement to the API and will reduce the boilerplate associated with converting back-and-forth between `Interval` and `BedRecord`.

Should a warning be logged (or an error raised) if the user attempts to query an `OverlapDetector` that was constructed from stranded interval-like objects with an unstranded query interval? (Or vice versa?)

File: pybedlite/overlap_detector.py [hunk 12]
```diff
@@ -235,7 +308,7 @@ def add_all(self, intervals: Iterable[Interval]) -> None:
         for interval in intervals:
             self.add(interval)
 
-    def overlaps_any(self, interval: Interval) -> bool:
+    def overlaps_any(self, interval: GenomicSpan) -> bool:
```
[the remark this answers is not in the document]
Comment (msto) [reply pybedlite/overlap_detector.py:311]: Or does this relate to my question in the top-level comment - are we deliberately supporting any `GenomicSpan` as input, regardless of the parameterized type of the detector?

[the remark this answers is not in the document]
Comment (clintval) [reply pybedlite/overlap_detector.py:311]: Yes, we're deliberately allowing you to query the overlap detector with _**other**_ features that could be different than the type stored inside the overlap detector. So, you could store BED records in the overlap detector but query for overlapping BED records using just a plain old interval.

File: pybedlite/overlap_detector.py [hunk 13]
```diff
@@ -56,6 +80,35 @@
...
+
+class StrandedGenomicSpan(GenomicSpan, Protocol):
+    """
+    A structural type for a stranded span on a reference sequence with zero-based open-ended
+    coordinates.
+    """
+
+    @property
+    def negative(self) -> bool:
+        """True if the interval is on the negative strand, False otherwise"""
```
[the remark this answers is not in the document]
Comment (clintval) [reply pybedlite/overlap_detector.py:109]: We considered this and could still go in this direction.

The benefit of having `negative` on `BedRecord` is it requires no changes to `Interval` and one additional helper method on `BedRecord`. But if we want to use `is_negative` then we would add that as a helper to `BedRecord` and would also need to add a **_duplicate getter_** to `Interval` (it already has a `negative(): bool` field) making more API bloat. I suppose we could break the API fully and rename the fields, but I don't have a good grasp on how much downstream/user code would have to change and it seems like we're trying to avoid that as completely as possible.

File: pybedlite/overlap_detector.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (msto) [reply pybedlite/overlap_detector.py:311]: Got it. I think that is sensible and will be generally helpful in practice.

I am concerned about the behavior when mixing stranded and unstranded interval types.

e.g. suppose you construct an overlap detector from the following bed:

```bed
chr1  100  300  intervalA  .  +
chr1  200  400  intervalB  .  -
```

And query it with the interval `Interval("chr1", 150, 250)`.

if I understand correctly, I _think_ that the current implementation will only return `intervalA` as an overlap, but given that the user provided an unstranded interval, would they be expecting to receive both intervals?

File: pybedlite/overlap_detector.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (msto) [reply pybedlite/overlap_detector.py:109]: My 2c would be to change the API now (we're on v0.1.0, I don't think there are high expectations for API stability 🙂 )

I'd be in favor of 
1. renaming `Interval`'s `negative` attribute to `is_negative`, 
2. adding a `negative(): bool` property which raises a `DeprecationWarning` and returns the value of `is_negative`, for backwards compatibility
3. cycling the `negative` property out in a few releases.

File: pybedlite/overlap_detector.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (clintval) [reply pybedlite/overlap_detector.py:311]: The implementation doesn't actually behave that way. This predates me, but strand is only used for sorting in one very specific situation here:

https://github.com/fulcrumgenomics/pybedlite/blob/d922a35383e1f676094ffcda1680c49aaa8feb62/pybedlite/overlap_detector.py#L356-L364

I'm not sure of the motivation for this decision, as I would have preferred the OverlapDetector to just return what overlaps, and if I wanted to spend the extra CPU cycles to sort, I'm OK doing that myself as a user of the API based on an ordering of my choosing.

Check out the original docs here: https://github.com/fulcrumgenomics/pybedlite/pull/34#discussion_r1700232535

I'll take another pass at the docs to make it super clear what the negative property is being used for.

[the remark this answers is not in the document]
Comment (msto) [reply pybedlite/overlap_detector.py:311]: Got it. If we are expanding `OverlapDetector` to accommodate stranded interval types, I think we should minimally emphasize in the documentation that strand is not considered when determining overlap, and perhaps open an issue to support stranded overlap (analogous to `bedtools intersect -s`)

File: pybedlite/overlap_detector.py (hunk 11 above)
Comment (msto) [reply pybedlite/overlap_detector.py:241]: A little, yes.

I think the documentation in the latter case could be made clearer by 1) excluding the type variable from the list of classes defined in the module and 2) making it clear that `OverlapDetector` is generic (which as far as I can tell is not currently apparent in the signature or docstring).

e.g.

> `class pybedlite.overlap_detector.OverlapDetector[S](intervals: Iterable[S] | None = None)`
>    Detects and returns overlaps between a set of genomic regions and another genomic region.


fwiw I don't think the average Python user thinks about generics _at all_, and I think it would be preferable to use established conventions for generic typing

Comment (msto) [reply pybedlite/overlap_detector.py:241]: I guess what I am trying to say is that I find the signature `OverlapDetector(intervals: Iterable[GenericGenomicSpan])` to be less clear than `OverlapDetector[S](intervals: Iterable[S])`, because the former does not make it clear that `OverlapDetector` is actually generic

Comment (clintval) [reply pybedlite/overlap_detector.py:241]: > 1) excluding the type variable from the list of classes defined in the module

Because we don't have this project setup where `__init__.py` defines the namespace the only way I could achieve this in this PR is by making the type variable private with `_X` (or eliminating the docstring so Sphinx doesn't autodoc it, but that is not favorable), but then it would show up in the documented type signature of the OverlapDetector as a private type variable. So I think for now it should be public until the project structure changes.

> 2) making it clear that OverlapDetector is generic (which as far as I can tell is not currently apparent in the signature or docstring).

We don't use the phrase generic, but what the OverlapDetector can contain is documented. I can certainly enhance the docs further to use the phrase "generic" too.

https://github.com/fulcrumgenomics/pybedlite/blob/74f7d031f48036fd2aa2646b17d4f375669e36fe/pybedlite/overlap_detector.py#L251-L252

> I guess what I am trying to say is that I find the signature OverlapDetector(intervals: Iterable[GenericGenomicSpan]) to be less clear than OverlapDetector[S](intervals: Iterable[S]), because the former does not make it clear that OverlapDetector is actually generic

The former is more specifically `OverlapDetector[GenericGenomicSpan](intervals: Iterable[GenericGenomicSpan])` and the word "generic" does appear in the type variable name. At least on that point, I don't agree that is all that unclear.

Comment (clintval) [reply pybedlite/overlap_detector.py:241]: ^ I bet Sphinx will someday support Py3.12 syntax which would be the specific form:

- `OverlapDetector[GenericGenomicSpan](intervals: Iterable[GenericGenomicSpan])`

I doubt it is capable of that now, though.

Comment (nh13) [reply pybedlite/overlap_detector.py:241]: Let's use Span here...
```suggestion
Span = TypeVar("GenericGenomicSpan", bound=Union[GenomicSpan, StrandedGenomicSpan])
```

File: pybedlite/overlap_detector.py [hunk 14]
```diff
@@ -258,70 +339,101 @@ def overlaps_any(self, interval: Interval) -> bool:
...
-              The list of intervals in this  detector that enclose the query interval.
-              The intervals will be returned in ascending genomic order.
+        Returns:
+            The list of intervals in this detector that enclose the query interval. The intervals
+            will be returned sorted using the following sort keys:
+
+                * The interval's start
+                * The interval's end
+                * The interval's strand (if defined)
+                * The interval's reference sequence name
```
Comment (nh13) [inline pybedlite/overlap_detector.py:402]: Do you think these clarifications help?  If so, can you add them everywhere?
```suggestion
                * The interval's start, ascending
                * The interval's end, ascending
                * The interval's strand, positive and negative (assumed to be positive if not defined)
                * The interval's reference sequence name (lexicographically)
```

File: pybedlite/tests/test_overlap_detector.py [hunk 15]
```diff
@@ -216,6 +219,144 @@ def test_construction_from_ucsc_with_strand(strand: str) -> None:
...
+        one_based_protocol,
+        zero_based_unstranded,
+        zero_based_stranded,
+    ]
+    assert detector.get_overlaps(Interval("chr1", 61, 62)) == [
+        zero_based_unstranded,
+        zero_based_stranded,
+    ]
+    assert detector.get_overlaps(Interval("chr1", 78, 79)) == [zero_based_stranded]
+    assert detector.get_overlaps(Interval("chr1", 80, 81)) == []
```
Comment (nh13) [inline pybedlite/tests/test_overlap_detector.py:337]: somewhere here I'd like to test (via mypy) that when we type hint the return types, mypy is ok with it.  Something like:

```python
    retvals: list[AllKinds] = detector.get_overlaps(Interval("chr1", 78, 79))

    detector: OverlapDetector[OneBasedProtocol]  = OverlapDetector(one_based_protocol)
    retvals: list[ OneBasedProtocol]  = detector.get_overlaps(one_based_protocol)
    assert retvals == [one_based_protocol]
```

File: pybedlite/overlap_detector.py (hunk 11 above)
Comment (msto) [reply pybedlite/overlap_detector.py:241]: That is worse IMO. 

It creates the implication that there exists a hierarchy `Span > GenomicSpan > StrandedGenomicSpan`, which can only increase confusion.

Given the voiced preference for a descriptive name for type variables, I would be fine leaving as `GenericGenomicSpan` or renaming to `GenomicSpanType` (slight preference for the latter, but we will defer to you)

File: pybedlite/overlap_detector.py (hunk 5 above)
Comment (nh13) [reply pybedlite/overlap_detector.py:290]: As not to pay the performance here, can we move this to where `set` is called and wrap it with a `try`/`except` for a `TypeError` that provides this exception message?  I think that's here:

https://github.com/fulcrumgenomics/pybedlite/blob/74f7d031f48036fd2aa2646b17d4f375669e36fe/pybedlite/overlap_detector.py#L366-L368

Comment (nh13) [reply pybedlite/overlap_detector.py:290]: Also, since `GenomicSpan` sub-classes `Hashable`, is this check redundant?

**Review by nh13 — CHANGES_REQUESTED** [review]
Small changes requested.  I'll take one more look and likely approve to merge.  Nice job!

File: pybedlite/overlap_detector.py [hunk 16]
```diff
@@ -56,6 +81,38 @@
 from pybedlite.bed_source import BedSource
 
 
+class GenomicSpan(Hashable, Protocol):
```
Comment (nh13) [inline pybedlite/overlap_detector.py:84]: Last minute question: should we get rid of `Genomic` which was introduced in this PR, since intervals can be relative to a transcriptome, amplicon, construct, plasmid, extracted sub-region, etc.  This would also shorten the names:
- `GenomicSpan` -> `Span`
- `StrandedGenomicSpan` -> `StrandedSpan`
- `GenericGenomicSpan` -> SpanType`

File: pybedlite/overlap_detector.py (hunk 11 above)
Comment (nh13) [reply pybedlite/overlap_detector.py:241]: Agreed, `SpanType` is clearer: https://github.com/fulcrumgenomics/pybedlite/pull/34/files#r1700567057

File: pybedlite/overlap_detector.py (hunk 5 above)
Comment (clintval) [reply pybedlite/overlap_detector.py:290]: > Also, since `GenomicSpan` sub-classes `Hashable`, is this check redundant?

TLDR: with this, we have 1 check at the type-level and 1 check at the implementation-level; both serve a purpose

Having `GenomicSpan` subclass `Hashable` ensures _**type safety**_, so MyPy will catch those misuses.

However, MyPy doesn't run at the _**implementation**_-level so that's where this check come in.

I do think both are important for a good user experience.

---

> As not to pay the performance here, can we move this to where `set` is called and wrap it with a `try`/`except` for a `TypeError` that provides this exception message? I think that's here:
> 
> https://github.com/fulcrumgenomics/pybedlite/blob/74f7d031f48036fd2aa2646b17d4f375669e36fe/pybedlite/overlap_detector.py#L366-L368

This is a good idea If we commonly **_store_** many more features than we return by the methods of this class. This is likely to be the common use case, so I'm happy to make this change!

File: pybedlite/overlap_detector.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (clintval) [reply pybedlite/overlap_detector.py]: If we do this, then I would also bundle breaking the API for `refname`/`chrom` as the primary property for the reference sequence too. I recommend we defer this to a future PR, though, and break everything all at once with the deprecation warnings as you recommend.

**Review by msto — APPROVED** [review]

**Review by nh13 — APPROVED** [review]
Nice work!
